### PR TITLE
U629/t647 cluster per player remake

### PR DESCRIPTION
**Pull request by tofb15**
*Metaballs is now clustered and rendered by player.
*Reworked accelleration Structures to allow multiple AABBs in the shaders.
*Changed the stepping algorithm in the metaball shader to make it faster
### NEW
All Players have their own AABB for their metaballs
AABB_new

### OLD
All Players share one AABB for metaballs
image

### CODE REVIEW

File: Sail/src/API/DX12/dxr/DXRBase.h [hunk 1]
```diff
@@ -153,7 +163,7 @@ class DXRBase final : public EventReceiver {
 	std::vector<MeshHandles> m_rtMeshHandles[2];
 	// Metaballs
 	std::vector<ID3D12Resource1*> m_metaballPositions_srv;
-	UINT m_metaballsToRender;
+	//UINT m_metaballsToRender;
```
Comment (whalemane) [inline Sail/src/API/DX12/dxr/DXRBase.h]: remove this?

File: Sail/src/API/DX12/renderer/DX12RaytracingRenderer.cpp [hunk 2]
```diff
@@ -63,53 +63,76 @@ void DX12RaytracingRenderer::present(PostProcessPipeline* postProcessPipeline, R
...
+	std::vector<int> groupsToRemove;
+	for (auto& group : m_metaballGroups) {
+		if (gpuGroupIndex >= MAX_NUM_METABALL_GROUPS || nMetaballs >= MAX_NUM_METABALLS) {
+			group.second.balls.clear();
+			group.second.gpuGroupStartOffset = 0;
+			group.second.index = -1;
+			groupsToRemove.push_back(group.first);
+			continue;
+		}
+		//std::reverse(group.second.balls.begin(), group.second.balls.end());
```
Comment (whalemane) [inline Sail/src/API/DX12/renderer/DX12RaytracingRenderer.cpp]: remove this?

**Comment by tofb15** [comment]
Crashes when a player dies. Probably related to the particle system but will look in to it.

**Comment by tofb15** [comment]
particle system fixed

**Pull request merged**